awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -105,32 +105,49 @@ def get_one(self, endpoint, *args, **kwargs):
 
         return response['json']['results'][0]
 
-    def resolve_name_to_id(self, endpoint, name_or_id):
-        # Try to resolve the object by name
+    def get_one_by_name_or_id(self, endpoint, name_or_id):
         name_field = 'name'
         if endpoint == 'users':
             name_field = 'username'
 
-        response = self.get_endpoint(endpoint, **{'data': {name_field: name_or_id}})
-        if response['status_code'] == 400:
-            self.fail_json(msg="Unable to try and resolve {0} for {1} : {2}".format(endpoint, name_or_id, response['json']['detail']))
+        query_params = {'or__{0}'.format(name_field): name_or_id}
+        try:
+            query_params['or__id'] = int(name_or_id)
+        except ValueError:
+            # If we get a value error, then we didn't have an integer so we can just pass and fall down to the fail
+            pass
+
+        response = self.get_endpoint(endpoint, **{'data': query_params})
+        if response['status_code'] != 200:
+            self.fail_json(
+                msg="Failed to query endpoint {0} for {1} {2} ({3}), see results".format(endpoint, name_field, name_or_id, response['status_code']),
+                resuls=response
+            )
 
         if response['json']['count'] == 1:
-            return response['json']['results'][0]['id']
+            return response['json']['results'][0]
+        elif response['json']['count'] > 1:
+            object_to_return = None
+            for tower_object in response['json']['results']:
+                # Name takes priority, so we match on name first
+                # If we already have an object to return and it has the right name, we need to fail
+                if object_to_return and object_to_return['name'] == name_or_id:
+                    self.fail_json(msg="The requested name or id was ambiguous and resulted in too many items")
+                # If we found an object ID with this name (which has to be unique) and we don't already have
+                # something to return, select this object for now
+                elif tower_object['id'] == name_or_id and not object_to_return:
+                    object_to_return = tower_object
```

Comment:
I still think there's something problematic here. I have drafted a test case where I'm creating a new team, and I give its related Organization as "1". Now there are 2 organizations present in the database:

 - Organization with name "foo" and id of 1
 - Organization with name "1" and id of 2

I give the related organization as `organization: 1`. This integer gets converted to a string by Ansible and comes in here as `name_or_id`. If it hits the org id=1 first, then that gets assigned to `object_to_return`. Thus, it fails on the 2nd item. However, if it hits the org id=2 first, then it does not fail. That's not good.